### PR TITLE
BEAM-2575 ApexRunner doesn't emit watermarks for additional outputs

### PR DESCRIPTION
**Pull request by tweise**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
R: @reuvenlax @kennknowles

### CODE REVIEW

File: runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java [hunk 1]
```diff
@@ -359,21 +359,28 @@ private void processWatermark(ApexStreamTuple.WatermarkTuple<?> mark) {
...
+      outputWatermark(ApexStreamTuple.WatermarkTuple.of(currentOutputWatermark));
+    }
+  }
+
+  private void outputWatermark(ApexStreamTuple.WatermarkTuple<?> mark) {
+    if (traceTuples) {
+      LOG.debug("\nemitting {}\n", mark);
+    }
+    output.emit(mark);
+    if (!additionalOutputPortMapping.isEmpty()) {
```
Comment (reuvenlax) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java]: is this if statement redundant?

Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java]: For efficiency, don't want to call values() unless needed.

Comment (reuvenlax) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java]: That's ok, though worth noting that the call to values() doesn't really incur much cost. I.e. it doesn't materialize the value set, it simply returns a special iterator.

**Comment by reuvenlax** [comment]
looks good to me

File: runners/apex/src/test/java/org/apache/beam/runners/apex/examples/WordCountTest.java [hunk 2]
```diff
@@ -123,11 +123,15 @@ public void testWordCountExample() throws Exception {
     options.setInputFile(new File(inputFile).getAbsolutePath());
     String outputFilePrefix = "target/wordcountresult.txt";
     options.setOutput(outputFilePrefix);
-    WordCountTest.main(TestPipeline.convertToArgs(options));
 
     File outFile1 = new File(outputFilePrefix + "-00000-of-00002");
     File outFile2 = new File(outputFilePrefix + "-00001-of-00002");
-    Assert.assertTrue(outFile1.exists() && outFile2.exists());
+    Assert.assertTrue(!outFile1.exists() || outFile1.delete());
```
Comment (kennknowles) [inline runners/apex/src/test/java/org/apache/beam/runners/apex/examples/WordCountTest.java]: This catches the issue, but how about using `TemporaryFolder`?

Example: https://github.com/apache/beam/blob/084a5e8aedfdd53f3d76ad48b55a66a282881646/examples/java/src/test/java/org/apache/beam/examples/DebuggingWordCountTest.java#L36

**Comment by kennknowles** [comment]
@tweise this seems worth cherry-picking to the 2.1.0 release branch. What do you think?

**Comment by kennknowles** [comment]
LGTM, failures not related, definitely an improvement - comments could be addressed afterwards.

**Comment by tweise** [comment]
@kennknowles +1 for cherry-pick. Will incorporate the tmp folder next time I stop by that test.

**Pull request merged**